linera-rpc/src/cross_chain_message_queue.rs

```diff
@@ -0,0 +1,199 @@
+    nickname: String,
+    cross_chain_max_retries: u32,
+    cross_chain_retry_delay: Duration,
+    cross_chain_sender_delay: Duration,
+    cross_chain_sender_failure_rate: f32,
+    this_shard: ShardId,
+    mut receiver: mpsc::Receiver<(CrossChainRequest, ShardId)>,
+    handle_request: F,
+) where
+    F: Fn(ShardId, CrossChainRequest) -> G + Send + Clone + 'static,
+    G: Future<Output = anyhow::Result<()>>,
+{
+    let mut futures = futures::stream::FuturesUnordered::new();
+    let mut job_states: HashMap<QueueId, Job> = HashMap::new();
+
+    let run_task = |task: Task| async move { handle_request(task.shard_id, task.request).await };
+
+    let run_action = |action, queue, state: Job| async move {
+        linera_base::time::timer::sleep(cross_chain_sender_delay).await;
+
+        let to_shard = state.task.shard_id;
+
+        (
+            queue,
+            match action {
+                Action::Proceed { .. } => {
+                    if let Err(error) = run_task(state.task).await {
+                        warn!(
+                            nickname = state.nickname,
+                            %error,
+                            retry = state.retries,
+                            from_shard = this_shard,
+                            to_shard,
+                            "Failed to send cross-chain query",
+                        );
+
+                        Action::Retry
+                    } else {
+                        trace!(from_shard = this_shard, to_shard, "Sent cross-chain query",);
+
+                        Action::Proceed {
+                            id: state.id.wrapping_add(1),
+                        }
+                    }
+                }
+
+                Action::Retry => {
+                    linera_base::time::timer::sleep(cross_chain_retry_delay * state.retries).await;
+                    Action::Proceed { id: state.id }
+                }
+            },
+        )
+    };
+
+    loop {
+        tokio::select! {
+            Some((queue, action)) = futures.next() => {
+                let Entry::Occupied(mut state) = job_states.entry(queue) else {
+                    panic!("running job without state");
+                };
+
+                let remove = matches!(action, Action::Proceed { id } if state.get().id < id)
+                    || matches!(action, Action::Retry if state.get().retries >= cross_chain_max_retries);
+                if remove {
+                    state.remove();
+                    continue;
+                }
+
+                if let Action::Retry = action {
+                    state.get_mut().retries += 1
+                }
+
+                futures.push(run_action.clone()(action, queue, state.get().clone()));
+            }
+
+            request = receiver.next() => {
+                let Some((request, shard_id)) = request else { break };
+                #[cfg(with_metrics)]
+                CROSS_CHAIN_MESSAGE_TASKS.set(job_states.len() as i64);
```

Comment:
Why not track `futures.len()` instead? It's the one that tracks how many cross-chain messages are in-flight.

Reply:
In that case we should update the metric when the job is successfully removed from the `job_states`?

Reply:
That seems sensible, we could just move this outside the loop?